### ISSUE TITLE
Full validation throws null reference exception when nested templates are used

### ISSUE THREAD

**Issue by StephenWeatherford** [microsoft/vscode-azurearmtools#716]
This seems to repro with *any* template using a nested template, if you have a parameter file associated with it (i.e., full validation is on).

1) Load any template with a nested template/deployment, e.g:
```
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.2.3.4",
    "resources": [
        {
            "type": "Microsoft.Resources/deployments",
            "name": "nested",
            "apiVersion": "2017-05-10",
            "properties": {
                "mode": "Incremental",
                "template": {
                    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
                    "contentVersion": "1.0.0.0",
                    "resources": [
                    ]
                }

            }
        }
    ]
}
```
2) Context menu -> Select/Create Parameter File... -> New... -> only required
3) go back to the template file
!) validation error:
{
	"resource": "/Users/stephwe/repos/template-examples/nested-templates-experiment.json",
	"owner": "ARM Template Language Server diagnostics",
	"severity": 8,
	"message": "Template validation failed: Internal error: Validation threw an exception: One or more errors occurred. (Object reference not set to an instance of an object.)",
	"source": "arm-template (validation)",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 1
}

**Comment by jorgecotillo** [comment]
ack, will look into the bug and provide a response.

**Comment by KenKilty** [comment]
Related to this bug there are also additional validation errors with inner/outer scoping for parameters and variables. The messages from the linter are incorrect but the issue is the same as Stephen's discovery with scoping of nested templates.

A slightly more complex repro that illustrates additional related errors:
1. Template validation failed: Internal error: Validation threw an exception: The template parameter 'paramFromOuter' is not found. Please see https://aka.ms/arm-template/#parameters for usage details.
2. Undefined parameter reference: 'paramFromOuter'
3. Undefined parameter reference: 'variableFromOuter'
4. Undefined variable reference: 'innerVariable'

Repro:

`{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "outerParam": {
            "type": "string",
            "defaultValue": "paramFromOuter"
        }
    },
    "variables": {
        "outerVariable": "varFromOuter"
    },
    "resources": [
        {
            "name": "nestedTemplate1",
            "apiVersion": "2019-10-01",
            "type": "Microsoft.Resources/deployments",
            "properties": {
                "expressionEvaluationOptions": {
                    "scope": "inner"
                },
                "mode": "Incremental",
                "template": {
                    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
                    "contentVersion": "1.0.0.0",
                    "parameters": {
                        "paramFromOuter": {
                            "type": "string"
                        },
                        "variableFromOuter": {
                            "type": "string"
                        }
                    },
                    "variables": {
                        "innerVariable": "varFromInner"
                    },
                    "resources": [
                    ],
                    "outputs": {
                        "paramFromOuter": {
                            "type": "string",
                            "value": "[parameters('paramFromOuter')]"
                        },
                        "variableFromOuter": {
                            "type": "string",
                            "value": "[parameters('variableFromOuter')]"
                        },
                        "variableFromInner": {
                            "type": "string",
                            "value": "[variables('innerVariable')]"
                        }
                    }
                },
                "parameters": {
                    "paramFromOuter": {
                        "value": "[parameters('outerParam')]"
                    },
                    "variableFromOuter": {
                        "value": "[variables('outerVariable')]"
                    }
                }
            }
        }
    ],
    "outputs": {
    }
}`

**Comment by StephenWeatherford** [comment]
@KenKilty Thanks!

**Comment by StephenWeatherford** [comment]
@jorgecotillo Two repros:
1) https://github.com/StephenWeatherford/template-examples/blob/master/716.json and https://github.com/StephenWeatherford/template-examples/blob/master/716.parameters.json (simple)
2) https://github.com/StephenWeatherford/template-examples/blob/master/716b.json and https://github.com/StephenWeatherford/template-examples/blob/master/716b.parameters.json (more complex - KenKilty's repro above, gives a different error.)

**Comment by StephenWeatherford** [comment]
https://github.com/microsoft/vscode-azurearmtools/pull/806

**Comment by StephenWeatherford** [comment]
Verified, added regression test.

**Comment by StephenWeatherford** [comment]
Shipped with 0.11.0
